frontend/packages/ceph-storage-plugin/src/components/create-storage-system/external-storage/types.ts

```diff
@@ -0,0 +1,95 @@
+import { TFunction } from 'i18next';
+import { ExtensionK8sModel } from '@console/dynamic-plugin-sdk/src';
+import { K8sResourceKind } from '@console/internal/module/k8s';
+import { IP_FAMILY } from '../../../constants';
+
+/**
+ *  Configures a new external storage vendor to the Storage System Creation wizard.
+ */
+export type ExternalStorage = {
+  /** Display name of the external storage vendor. */
+  displayName: string;
+
+  /** The model referring the `apiGroup`,`apiVersion` and `kind` of the external storage vendor's CRD. */
+  model: ExtensionK8sModel;
+
+  /** A React Functional Component to input the connection details of the external storage vendor. */
+  Component: React.FunctionComponent<ExternalComponentProps<{}>>;
+
+  /**  Handler function to create external storage storage vendor CR or resources. */
+  createPayload: CreatePayload<{}>;
+
+  /**  Handler function to validate the storage class page in order to move to the next step of wizard */
+  canGoToNextStep: CanGoToNextStep<{}>;
+};
+
+/** Props for `ExternalStorage.Component` to input the connection details of the external storage vendor. */
+export type ExternalComponentProps<S extends ExternalState> = {
+  /** The state of the `ExternalStorage.Component`. */
+  formState: S;
+
+  /** The callback for setting the state of `ExternalStorage.Component` */
+  setFormState: (field: string, value: ExternalStateValues) => void;
```

Comment:
keys of the state object will always be `string`

Reply:
I am not sure about that AFAIK If `S` is an object. 
Example:
```
type S = {
   'field': string;
}
``` 
`keyof S` would give you `field` whereas `typeof keyof S` would give you `string`.

Reply:
The stet object will be like 
```
{
username: 'kubeadmin'
}
```
Here the `field` will always be of type "string"

Reply:
```
type S = {
username: string
}
```
`keyof S` would be `username`
```
const state = {
username: 'kubeadmin'
}
```
`keyof state` would be `never`.
Why are we comparing object and their type definitions. AFAIK S refers to a state.

Reply:
To clarfiy, S will always be an object right, and the keys of that object will always be string, so why if I get keyof S, it will be string. I will put a generic type of S. NO opinions.

Reply:
S will always be a type. Why would you pass an object to a typescript generic type?